### PR TITLE
feat: added omni xchain ui changes

### PR DESCRIPTION
**Pull request by percival7777**
*[GitHub keywords to close any associated issues](https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/)*

## Motivation

This PR adds ability to display omni xchain txs in the explorer

### Enhancements
*Things you added that don't break anything.  Regression tests for Bug Fixes count as Enhancements.*

### Bug Fixes
*Things you changed that fix bugs.  If a fixes a bug, but in so doing adds a new requirement, removes code, or requires a database reset and reindex, the breaking part of the change should be added to Incompatible Changes below also.*

### Incompatible Changes
*Things you broke while doing Enhancements and Bug Fixes.  Breaking changes include (1) adding new requirements and (2) removing code.  Renaming counts as (2) because a rename is a removal followed by an add.*

## Upgrading

*If you have any Incompatible Changes in the above Changelog, outline how users of prior versions can upgrade once this PR lands or when reviewers are testing locally.  A common upgrading step is "Database reset and re-index required".*

## Checklist for your Pull Request (PR)

  - [ ] I added an entry to `CHANGELOG.md` with this PR
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [ ] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

File: apps/block_scout_web/assets/js/omni.js [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+};
+
+const watch = async () => {
+
+  // update api url depending on blockscouts deployed environment
+  if (window.location.hostname.startsWith("testnet")) {
+    url = "testnet-1-xchain.explorer.omni.network";
+  } else if (window.location.hostname.startsWith("staging")) {
+    url = "staging-xchain.explorer.omni.network";
+  }
```
Comment (percival7777) [inline apps/block_scout_web/assets/js/omni.js]: These should match the cloudflare domains for xchain indexer api server in each env

File: apps/block_scout_web/assets/js/omni.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  const response = await fetch(url + "/checkOmniTxHashes", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({
+      txHashes: elements,
+    }),
+  });
+  if (response.status == 400) {
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: what if error is something else, 500 can happen most likely?

File: apps/block_scout_web/assets/js/omni.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  <img style="height: 2vw; display: inline-block;margin-top: -0.3vw" src="/images/favicon-32x32.png"> 
+  <div style="margin-top: 0.1vw;display: inline-block;font-weight: bolder;color: white;font-size: 1.2vw;padding-left: 0.5vw;">X-Chain Transaction Details</div></div>
+  <pre style="font-size: 1vw;display: block;margin-top: 0.5vw;font-weight: bolder;color: white;padding-left: 1vw;">
+  ${JSON.stringify(details, null, 4)}
+  </pre>`;
+  return div;
+};
+
+// expected return {txHash1:true, txHas2:true...}
+const checkTx = async (elements) => {
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: checkTxs seems like a better name
what is this checking?

File: apps/block_scout_web/assets/js/omni.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    return undefined;
+  }
+  return response.json();
+};
+
+const getTxDetails = async (omniTxHash) => {
+  // dummy response
+  const response = await fetch(url + `/omniTx/${omniTxHash}`);
+  const data = await response.json();
+  if (data.length === 0) {
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: why not return data as empty array (on the other hand, this should return details of single tx, or array)? can this endpoint return errors?

File: apps/block_scout_web/lib/block_scout_web/csp_header.ex [hunk 5]
```diff
@@ -19,7 +19,7 @@ defmodule BlockScoutWeb.CSPHeader do
 
     Controller.put_secure_browser_headers(conn, %{
       "content-security-policy" => "\
-        connect-src 'self' #{json_rpc_url} #{config[:mixpanel_url]} #{config[:amplitude_url]} #{websocket_endpoints(conn)} #{czilladx_url} #{trustwallet_url} #{walletconnect_urls};\
+        connect-src 'self' http://localhost:8888 https://testnet-1-xchain.explorer.omni.network https://staging-xchain.explorer.omni.network #{json_rpc_url} #{config[:mixpanel_url]} #{config[:amplitude_url]} #{websocket_endpoints(conn)} #{czilladx_url} #{trustwallet_url} #{walletconnect_urls};\
```
Comment (skosito) [inline apps/block_scout_web/lib/block_scout_web/csp_header.ex]: should this be parametrized as the rest?

File: apps/block_scout_web/assets/js/omni.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+const handletxHashLinks = async () => {
+  const elements = getAllTxHashElements();
+  if (elements.length == 0) {
+    return;
+  }
+  // check local cache
+  const unknownTxList = [];
+  for (let i of elements) {
+    if (i.parentNode.childElementCount <= 2 && txMap[i.innerHTML] === true) {
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: not sure how much we will maintain this, but i think things that are not clear from condition can be moved into a function with descriptive name, for example here is not so clear why childCount should be <=2 and what is txMap holding
but up to you

File: apps/block_scout_web/assets/js/omni.js [hunk 7]
```diff
@@ -0,0 +1,163 @@
+// local cache to re-use data
+let txMap = {};
+
+// default api url point to local server
+let url = "http://localhost:8888";
+
+const customMarker = () => {
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: is this xchain marker just adding xchain image or something else? maybe lets name it like that

File: apps/block_scout_web/assets/js/omni.js [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  const xChainTxs = await checkTx(unknownTxList);
+  if (xChainTxs === undefined) {
+    return;
+  }
+  for (let i of elements) {
+    if (xChainTxs[i.innerHTML] === true) {
+      txMap[i.innerHTML] = true;
+      console.log("from api " + i.innerHTML);
+      i.parentNode.appendChild(customMarker());
+    } else if (txMap[i.innerHTML] == undefined) {
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: ===

potentially not needed since txMap[..] is false in any condition if its undefined or false, but i guess this is more readable since undefined means it is not set and false means it is set to false, but maybe some code can be removed

File: apps/block_scout_web/assets/js/omni.js [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+      txMap[i.innerHTML] = false;
+    }
+  }
+};
+
+const displayTxDetails = async (txHash) => {
+  const elements = document.querySelectorAll(
+    '[class="card js-ad-dependant-mb-3"]'
+  );
+  if (elements.length != 1) {
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js:111]: if we have control over this maybe we can add id and getElementById but this is also ok

File: apps/block_scout_web/assets/js/omni.js (hunk 1 above)
Comment (skosito) [reply apps/block_scout_web/assets/js/omni.js]: can these be params somehow? if we change domain we have to change codebase

File: apps/block_scout_web/assets/js/omni.js [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    url = "testnet-1-xchain.explorer.omni.network";
+  } else if (window.location.hostname.startsWith("staging")) {
+    url = "staging-xchain.explorer.omni.network";
+  }
+
+  // handle for homepage
+  if (window.location.pathname === "/") {
+    repeatLoop(1500);
+  } else if (window.location.pathname.startsWith("/address/")) {
+    // every 5 second because there can be too many updates
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: comment says 5 but looks like 2, same below

File: apps/block_scout_web/assets/js/omni.js [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  // update api url depending on blockscouts deployed environment
+  if (window.location.hostname.startsWith("testnet")) {
+    url = "testnet-1-xchain.explorer.omni.network";
+  } else if (window.location.hostname.startsWith("staging")) {
+    url = "staging-xchain.explorer.omni.network";
+  }
+
+  // handle for homepage
+  if (window.location.pathname === "/") {
+    repeatLoop(1500);
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js:144]: is repeatLoop mechanism how blockscout ui is doing api calls, or this is custom thing for us? is there a difference in rate in which the rest of blockscout ui is polling for data

File: apps/block_scout_web/assets/js/omni.js [hunk 12]
```diff
@@ -0,0 +1,163 @@
+// local cache to re-use data
+let txMap = {};
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: you probably checked, but there is no way to make some sort of frontend component(s) with some ui framework blockscout is using, we have to do DOM manipulation manually? it seems we are doing some dom manipulation in every loop iteration, which some framework would improve a lot

File: apps/block_scout_web/assets/js/omni.js [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+  const elements = getAllTxHashElements();
+  if (elements.length == 0) {
+    return;
+  }
+  // check local cache
+  const unknownTxList = [];
+  for (let i of elements) {
+    if (i.parentNode.childElementCount <= 2 && txMap[i.innerHTML] === true) {
+      console.log("from cache " + i.innerHTML);
+      i.parentNode.appendChild(customMarker());
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: if txMap is already set in local cache, do we need to basically re-attach and rerender this every time? can we call txMap xChainsTxsMap if that is more suitable name?

File: apps/block_scout_web/assets/js/omni.js [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    } else if (txMap[i.innerHTML] === undefined) {
+      unknownTxList.push(i.innerHTML);
+    }
+  }
+  if (unknownTxList.length == 0) {
+    return;
+  }
+  const xChainTxs = await checkTx(unknownTxList);
```
Comment (skosito) [inline apps/block_scout_web/assets/js/omni.js]: lets add more descriptive name for checkTx, saying what is exaclty  being checked

**Review by skosito — APPROVED** [review]
nice, left some comments around, please check what makes sense before merging

File: apps/block_scout_web/assets/js/omni.js (hunk 2 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: The 400 is handled because that's business logic if there's a different error the UI would log it in console as that's not expected.

File: apps/block_scout_web/assets/js/omni.js (hunk 3 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: which txs in the list are xchain

File: apps/block_scout_web/assets/js/omni.js (hunk 4 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: It can returns array, the undefined here is just for the ui to understand that this is not a xchain tx so no ui changes will be made.

File: apps/block_scout_web/lib/block_scout_web/csp_header.ex (hunk 5 above)
Comment (percival7777) [reply apps/block_scout_web/lib/block_scout_web/csp_header.ex]: didn't find any need for that because these are static values and will not change dynamically.

File: apps/block_scout_web/assets/js/omni.js (hunk 6 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: This is a patch, it's mostly a hacky solution and can break if blockscout plans to change any ui component. Our goal should be to create an UI from scratch for the xchain indexer.

File: apps/block_scout_web/assets/js/omni.js (hunk 7 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js:7]: as the name suggests it gives the custom marker for xchain on txs, it's more than an image.

File: apps/block_scout_web/assets/js/omni.js (hunk 12 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: blokscout doesn't use any ui framework, it's server side rendering with bare minimal dynamic js. I am not an ui expert but this is the best solution I could find that can be implemented in the given timeline.

File: apps/block_scout_web/assets/js/omni.js (hunk 1 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: Nothing is dynamic here all these are static files running on the browser so cannot be parameterized

File: apps/block_scout_web/assets/js/omni.js (hunk 11 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js:144]: Blockscout is server side rendered mostly and uses websockets for pushing live data from backend to frontend. Our changes are patch work changes made to add extra info in the UI so it's completely separate from blockscout.

File: apps/block_scout_web/assets/js/omni.js (hunk 9 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js:111]: I don't think we can make that change without changing the server code for blockscout, this works for now until blockscout changes their ui components.

File: apps/block_scout_web/assets/js/omni.js (hunk 8 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: Don't think txMap[<missing_key>] will give false as the return, maybe in typescript with strict types but here it will return undefined so I set it to false manually. We need to do this because we need to identify items not yet seen(missing from the map) and items that have been already visited/noticed.

File: apps/block_scout_web/assets/js/omni.js (hunk 7 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: will rename to xchainMarker()

File: apps/block_scout_web/assets/js/omni.js (hunk 13 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: blockscout ui re-renders every time it gets a live update so we lose all the markets hence we need to re-render these for any new or old cached txs. it is not a map of xChain txs but rather a map of all txs we have seen, true indicates it's an xchain tx while false indicates it's a normal tx

File: apps/block_scout_web/assets/js/omni.js (hunk 9 above)
Comment (kevinhalliday) [reply apps/block_scout_web/assets/js/omni.js:111]: agree with stefan here. would prefer an explicit selector. either an id, if it's unique element perpage. or a data-selector. we can add these by updating blockscout templates [here](https://github.com/omni-network/blockscout/blob/feat/omni_changes/apps/block_scout_web/lib/block_scout_web/templates/transaction/overview.html.eex#L34). or we could add a placeholder element to that template with id="omni-xchain-details" or something like that

File: apps/block_scout_web/assets/js/omni.js (hunk 11 above)
Comment (kevinhalliday) [reply apps/block_scout_web/assets/js/omni.js:144]: okay with this for now. some thoughts.

- can use setInterval instead of setTimeout if you'd like. also, it shouldn't be an issue because we only start a loop at page load, but in case that changes, we should clear our timeouts / intervals when they're no longer needed. to avoid running loops in the background that we don't need
- rather than using setInterval / setTimeout, we might consider a [mutation observer](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver), to watch for changes in say, a list of transaction. and then react to new elements accordingly. should make our reaction to new txs a bit smoother
- at some point would be nice to inject our omni xchain details into the websocket new_transactions channels, to avoid a delay in displaying xchain markers

File: apps/block_scout_web/assets/js/omni.js (hunk 8 above)
Comment (kevinhalliday) [reply apps/block_scout_web/assets/js/omni.js]: is innerHtml always just the transaction hash? anyway we can get this more explictly? maybe by using a data tag on the elements? i believe blockscout adds a `data-identifier-hash="..."` to transaction cards. could use that

also, rather than using inmemory cache, you could add a data tag to elements you've already checked. then, when you query the dom, you can use a selector that filters out elements with that tag. this would speed up dom query. you might need to add an "unchecked" tag in the template

File: apps/block_scout_web/assets/js/omni.js (hunk 12 above)
Comment (kevinhalliday) [reply apps/block_scout_web/assets/js/omni.js]: perhaps a more explict name here. I wasn't sure what this was for till I read usage

File: apps/block_scout_web/assets/js/omni.js [hunk 15]
```diff
@@ -0,0 +1,162 @@
+// local cache to re-use data
+let txMap = {};
+
+// default api url point to local server
+let url = "http://localhost:8888";
```
Comment (kevinhalliday) [inline apps/block_scout_web/assets/js/omni.js]: `indexerUrl`?

File: apps/block_scout_web/assets/js/omni.js (hunk 6 above)
Comment (kevinhalliday) [reply apps/block_scout_web/assets/js/omni.js]: then perhaps a comment to explain the childElement bit?

File: apps/block_scout_web/assets/js/omni.js [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+const customXChainDataView = (details) => {
+  const div = document.createElement("div");
+  div.style =
+    "height: inherit;background-color: rgb(130, 88, 205);padding: 2.5vw 1.5vw 0vw 1.5vw;";
+  div.innerHTML = `
+  <div>
+  <img style="height: 2vw; display: inline-block;margin-top: -0.3vw" src="/images/favicon-32x32.png"> 
+  <div style="margin-top: 0.1vw;display: inline-block;font-weight: bolder;color: white;font-size: 1.2vw;padding-left: 0.5vw;">X-Chain Transaction Details</div></div>
+  <pre style="font-size: 1vw;display: block;margin-top: 0.5vw;font-weight: bolder;color: white;padding-left: 1vw;">
+  ${JSON.stringify(details, null, 4)}
```
Comment (kevinhalliday) [inline apps/block_scout_web/assets/js/omni.js:25]: we'll probably want to make this a bit prettier at some point. but it does have a nice software engineer vibe

File: apps/block_scout_web/assets/js/omni.js (hunk 2 above)
Comment (kevinhalliday) [reply apps/block_scout_web/assets/js/omni.js]: might as well throw the whole fetch in a try / except block. so we catch anything unexpected. like a network error

**Review by kevinhalliday — APPROVED** [review]
Nice work figuring this out. I know it was funky, given how blockscout implements their UI. The blanket approach here just watching the DOM for new tx cards is nice and direct. Simpler than getting into the internals of their webbsocket channels / redux store.

Some bits  do feel a little "patchy", as you said, but that's okay for now. We can revisit later.

Few comments, some repeated from below

- blockscout does use jquery, so we can use that. it helps clean up code that's interacting with the DOM
- let's use explicit selectors whenever we can. and maybe use data tags on nodes. let the DOM query utilities do some of the work for us
- let's make a ticket to try our replacing our `repeatLoop` function with a mutation observer. that way, we can react to just new nodes as soon as they are added. one potential issue here is that our indexer may be behind blockscouts, so if we a new tx card is added, and we query our indexer, but receive no result. we'd a way to keep checking until we know it is or is not an xchain tx
- at some point it would be nice to get our xchain details into the transaction websocket channels, so we have everything we need to render a card when that event is received. or, if that's infeasible, we can use their websocket events to trigger checks against our indexer. though this would require getting into the internals of blockscouts redux, which 
- for transaction page, it would be nice to get all our xchain details in at server render time. but this is not a priority

And some stylstic things, if anything to keep my linter quiet. Also good to just get into best practice. Even if they seem silly at times.
- basically always use `===` over `==`
- use `const` over `let`

File: apps/block_scout_web/assets/js/omni.js [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+    setTimeout(repeatLoop, timeMS);
+  }
+};
+
+const watch = async () => {
+  // update api url depending on blockscouts deployed environment
+  if (window.location.hostname.startsWith("testnet")) {
+    url = "https://testnet-1-xchain.explorer.omni.network";
+  } else if (window.location.hostname.startsWith("staging")) {
+    url = "https://staging-xchain.explorer.omni.network";
```
Comment (ttarsi) [inline apps/block_scout_web/assets/js/omni.js]: we don't need to append "-xchain"

File: apps/block_scout_web/assets/js/omni.js (hunk 11 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js:144]: 1. I use setTimeout because I use the pattern where we run the next iteration wait only when the previous one completes, setInterval just fires automatically without waiting for the actual function to finish as far as I have seen.

2. Not using mutation observer because if you see our testnet env the incoming tx rate is very high we don't want to make an api call for each and every single new incoming tx. The interval approach batches a few txs and reduces the load on the servers significantly as the expense of a small delay in the UI which I think is probably fine. We are not going to settle on this patch work ui and it'll probably never be the solution. We will have to build a new UI from scratch using something like react leveraging the blockscout apis and xchain indexer apis in more optimized way.

File: apps/block_scout_web/assets/js/omni.js (hunk 8 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js]: No it's not possible because blosckscout re-renders the UI every time it gets a new tx so all modifications are removed. That's why I need the in memory cache.

File: apps/block_scout_web/assets/js/omni.js (hunk 9 above)
Comment (percival7777) [reply apps/block_scout_web/assets/js/omni.js:111]: I am pretty sure this tag I use is unique for the txHash, the check is just there make sure that assumption doesn't change with a future update or something and break something else.

**Comment by percival7777** [comment]
Merging it for now to stabilize staging, updated terraform to work with static image for testnet while staging will use the new builds. Most major changes requested have been done any other improvements we can take them up later.
